backend/asm_targets/asm_directives.ml

```diff
@@ -0,0 +1,339 @@
+      in
+      if not first_occurrence then new_line ();
+      D.section names flags args;
+      if first_occurrence then define_label (Asm_label.for_section section)
+
+  let initialize () =
+    cached_strings := Cached_string.Map.empty;
+    sections_seen := [];
+    temp_var_counter := 0;
+    current_dwarf_section_ref := None;
+    (* Forward label references are illegal in GAS.
+       To avoid it, emit the beginning of all dwarf sections in advance. *)
+    begin
+      if is_gas () then
+        List.iter switch_to_section (Asm_section.dwarf_sections_in_order ())
+    end;
+    (* Stop dsymutil complaining about empty __debug_line sections (produces
+       bogus error "line table parameters mismatch") by making sure such
+       sections are never empty. *)
+    let file_num = A.get_file_num "none" in
+    loc ~file_num ~line:1 ~col:1;
+    D.text ()
+
+  let with_comment f ?comment x =
+    Option.iter D.comment comment;
+    f x
+
+  let ( >> ) f g x = g (f x)
+
+  let int8 =
+    with_comment (Int8.to_int >> Int64.of_int >> D.const_int64 >> D.byte)
+
+  let int16 =
+    with_comment (Int16.to_int >> Int64.of_int >> D.const_int64 >> D.word)
+
+  let int32 = with_comment (Int64.of_int32 >> D.const_int64 >> D.long)
+
+  let int64 = with_comment (D.const_int64 >> D.qword)
+
+  let uint8 =
+    with_comment (Uint8.to_int >> Int64.of_int >> D.const_int64 >> D.byte)
+
+  let uint16 =
+    with_comment (Uint16.to_int >> Int64.of_int >> D.const_int64 >> D.word)
+
+  let uint32 = with_comment (Uint32.to_int64 >> D.const_int64 >> D.long)
+
+  let uint64 = with_comment (Uint64.to_int64 >> D.const_int64 >> D.qword)
+
+  let targetint ?comment num =
+    match Targetint.repr num with
+    | Int32 n -> int32 ?comment n
+    | Int64 n -> int64 ?comment n
+
+  let uleb128 = with_comment (Uint64.to_int64 >> D.const_int64 >> D.uleb128)
+
+  let sleb128 = with_comment (D.const_int64 >> D.sleb128)
+
+  let assert_string_has_no_null_bytes s =
+    assert (not (String.contains s '\x00'))
+
+  let string =
+    with_comment (fun str ->
+      assert_string_has_no_null_bytes str;
+      D.bytes str)
+
+  let cache_string ?comment section str =
+    assert_string_has_no_null_bytes str;
+    let cached : Cached_string.t = { section; str; comment } in
+    match Cached_string.Map.find cached !cached_strings with
+    | label -> label
+    | exception Not_found ->
+      let label = Asm_label.create section in
+      cached_strings := Cached_string.Map.add cached label !cached_strings;
+      label
+
+  let emit_cached_strings () =
+    let old_dwarf_section = !current_dwarf_section_ref in
+    Cached_string.Map.iter
+      (fun { section; str; comment } label_name ->
+        switch_to_section section;
+        define_label label_name;
+        string ?comment str;
+        int8 Int8.zero)
+      !cached_strings;
+    cached_strings := Cached_string.Map.empty;
+    current_dwarf_section_ref := old_dwarf_section
```

Comment:
Should this be "switch_to_section old_dwarf_section" instead, to keep  the assembly file in sync?

Reply:
Fixed